src/librustc_lint/types.rs

```diff
@@ -676,3 +684,63 @@ impl LateLintPass for ImproperCTypes {
         }
     }
 }
+
+pub struct VariantSizeDifferences;
+
+impl LintPass for VariantSizeDifferences {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(VARIANT_SIZE_DIFFERENCES)
+    }
+}
+
+impl LateLintPass for VariantSizeDifferences {
+    fn check_item(&mut self, cx: &LateContext, it: &hir::Item) {
+        if let hir::ItemEnum(ref enum_definition, ref gens) = it.node {
+            if gens.ty_params.is_empty() {  // sizes only make sense for non-generic types
+                let mut sizes = vec![];
+                let t = cx.tcx.node_id_to_type(it.id);
+                let layout = cx.tcx.normalizing_infer_ctxt(ProjectionMode::Any).enter(|infcx| {
+                    t.layout(&infcx).unwrap_or_else(|e| {
+                        bug!("failed to get layout for `{}`: {}", t, e)
+                    })
+                });
+
+                if let Layout::General { ref variants, ref size, discr, .. } = *layout {
+                    let discr_size = Primitive::Int(discr).size(&cx.tcx.data_layout).bytes();
+
+                    debug!("enum `{}` is {} bytes large", t, size.bytes());
+
+                    for (variant, variant_layout) in enum_definition.variants.iter().zip(variants) {
+                        // Subtract the size of the enum discriminant
+                        let bytes = variant_layout.min_size().bytes().saturating_sub(discr_size);
+                        sizes.push(bytes);
```

Comment:
This could maybe be `.map(|(variant, variant_layout)| {...})` instead? That is, `sizes` doesn't seem to need to be collected into a `Vec`.